### ISSUE TITLE
Difficulty reproducing ConceptNet scores

### ISSUE THREAD

**Issue by chiayewken** [atcbosselut/comet-commonsense#3]
I'm having some difficulty reproducing the conceptnet accuracy scores, could you please point me in the right direction? These are the steps I took:
1. Download pretrained COMET model from https://drive.google.com/open?id=1FccEsYPUHnjzmX-Y5vjCBeyRt1pLo8FB
2. Run generations with greedy decoding on conceptnet test set
3. Use  Bilinear AVG model by Li et al. (2016) to score generated tuples and threshold by 0.5 to get accuracy

However, because I achieved an accuracy of 80.04%, which is much lower than the 95.25% reported in your Table, I'm sure I must be doing something wrong! The (short) code I used to get the scores is here: https://colab.research.google.com/drive/10oaX-_1qS75xrgbm67MDkRy_hOHEJA0s

**Comment by atcbosselut** [comment]
How many examples from the test set are you generating? The test set has 2400 examples, but we only generate the first 1200 of them in our work since the last 1200 are considered to be False (and the way they're generated leads to strange subject-relation pairs -- e.g., "person" + "HasPrerequisite"). 

I don't know if the classifier's performance would drop so drastically based on that detail alone, but I think you're generating for the full 2400 item test set with your script. Our generations are only for the first 1200. Let me know whether that covers the difference.

**Comment by chiayewken** [comment]
Originally the 80.04% score was evaluated on 2400 examples from the test set. When I changed to use only the first 1200 examples, the score is 92.42% which is much closer, thank you for your pointer! Are there any other details which might help close the gap? I noticed that the scoring model from Li et el doesn't produce a valid score for certain relations or if it encounters and unknown token. How are the scores for those examples factored into the final score?

**Comment by chiayewken** [comment]
Alternatively, could you provide the same evaluation script that was used to generate examples and calculate the scores in the paper?

**Comment by atcbosselut** [comment]
Generation of sequences should happen during training at each epoch:

https://github.com/atcbosselut/comet-commonsense/blob/c8a28e36a6e11689af749e60ccc8f2ac02cb0d3b/src/train/train.py#L139

I've added the scripts for classifying tuples and updated the README so you can find the steps to download the pretrained classifier for Li et al., 2016

https://github.com/atcbosselut/comet-commonsense/commit/2366023143c30573693c22a8ea5f5ffefa4ee080

**Comment by chiayewken** [comment]
Thank you for the scripts, but it seems that "classify.sh" calls a "convert_conceptnet_generations_to_text.py" which is missing. I wrote a quick replacement script, followed these steps to achieve a test set result of "Accuracy @ 0.5: 0.934166666667":

1. Download pretrained COMET model from https://drive.google.com/open?id=1FccEsYPUHnjzmX-Y5vjCBeyRt1pLo8FB
2. Run "generate_conceptnet_beam_search.py" with beam=1 for greedy decoding and split="test"
3. Run "classify.sh" on the generated test.pickle

The reproducibility gap is smaller but still present, may I refer you to my notebook code for advice on how to close this gap? Unfortunately, I do not have the resources to train your model from scratch. Thank you for your help!

Notebook link: https://colab.research.google.com/drive/10oaX-_1qS75xrgbm67MDkRy_hOHEJA0s

**Comment by atcbosselut** [comment]
Hi Chia Yew, I re-ran the greedy decoding for the provided pre-trained model and got the same number you did (~93.42%). Numbers can vary across different trainings of the model and the released pre-trained model is actually *not* the same one that was used for the numbers in the paper. When I cleaned up the code base for public release, I retrained a model.

I can give you the raw generations from the model from the paper if you'd like to test those against the results of the paper, but because the released pre-trained model is from a later version, the results wouldn't be exactly the same. In fact, due to non-determinism, if you were to re-train a model on your machine, you'd probably not get the same results as the current pretrained model either.

I want to also note, though, that the Classifier scores aren't a perfect metric. Since it's a model-based metric, the uncertainty of the evaluation model plays a role in how to interpret those numbers. The evaluation classifier will have false negatives (and possible false positives). While I would believe that a model that scores 10%+ higher on the classifier score is probably better than another, I'd expect that a 2% difference (what you're seeing) on a 1200 example set might fall safely within the margin of error of the classifier score, so I wouldn't worry too much about the difference. Let me know if I can get you anything else, but for now I'll close the issue!
